File: tests/fixtures/python/good_handler.py
"""A correct durable handler — the checker must report nothing on this file.

Structurally close to the bad fixture on purpose: same work, same shape, but
every nondeterministic operation is checkpointed. If a rule fires here it is a
false positive, which for a linter is the failure that gets it uninstalled.
"""

import time
import uuid

import boto3
from aws_durable_execution_sdk_python import DurableContext, durable_execution

table = boto3.resource("dynamodb").Table("orders")


@durable_execution
def lambda_handler(event, context: DurableContext):
    run_id = event["run_id"]

    # Clock and identity are checkpointed, so replay reuses the original values.
    started_at = context.step(lambda _: time.time(), name="started-at")
    correlation_id = context.step(lambda _: str(uuid.uuid4()), name="correlation-id")

    # The decision is checkpointed, so control flow is stable across replay.
    shift = context.step(
        lambda _: "morning" if time.localtime().tm_hour < 12 else "afternoon",
        name="pick-shift",
    )

    # Values come back through return values and are assigned in the handler,
    # where replay rebuilds them from checkpointed results.
    receipts = []
    receipt_id = context.step(
        lambda _: table.put_item(Item={"id": run_id, "started": started_at}),
        name="save-order",
    )
    receipts.append(receipt_id)

    # Branching on checkpointed data is fine.
    if shift == "morning":
        context.step(lambda _: "morning-work", name="morning-work")
    else:
        context.step(lambda _: "afternoon-work", name="afternoon-work")

    # Loop suffixes derived from a stable index, not from a clock.
    for index, item in enumerate(event.get("items", [])):
        context.step(lambda _, i=item: str(i), name=f"item-{index}")

    return {
        "run_id": run_id,
        "correlation_id": correlation_id,
        "receipts": receipts,
    }

File: tests/fixtures/rust/good_handler.rs
//! A correct durable handler -- the checker must report nothing here.
//!
//! Same work as the bad fixture, every nondeterministic operation checkpointed.

use chrono::Utc;
use durable_lambda_core::context::DurableContext;
use uuid::Uuid;

#[durable_execution]
async fn handler(event: Value, mut ctx: DurableContext) -> Result<Value, Error> {
    // Clock and identity are checkpointed, so replay reuses the values.
    let started = ctx
        .step("started_at", || async { Ok::<_, String>(Utc::now().to_rfc3339()) })
        .await?;

    let correlation = ctx
        .step("correlation_id", || async { Ok::<_, String>(Uuid::new_v4().to_string()) })
        .await?;

    // The decision itself is checkpointed, so control flow is stable.
    let shift = ctx
        .step("pick_shift", || async {
            Ok::<_, String>(if Utc::now().hour() < 12 { "morning" } else { "afternoon" })
        })
        .await?;

    // I/O belongs inside a step -- that is what a step is for.
    let receipt = ctx
        .step("save_order", || async {
            let body = std::fs::read_to_string("/tmp/order.json")?;
            Ok::<_, String>(body)
        })
        .await?;

    // Branching on checkpointed data is fine.
    if shift == "morning" {
        ctx.step("morning_work", || async { Ok::<_, String>("am") }).await?;
    }

    // A loop suffix from a stable index, not from a clock.
    for index in 0..3 {
        ctx.step(&format!("item-{}", index), || async { Ok::<_, String>(index) })
            .await?;
    }

    Ok(json!({ "started": started, "correlation": correlation, "receipt": receipt }))
}
